phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1308,6 +1345,19 @@ public void postBatchMutateIndispensably(ObserverContext<RegionCoprocessorEnviro
       try {
           if (success) {
               context.currentPhase = BatchMutatePhase.POST;
+              for (int i = 0; i < miniBatchOp.size(); i++) {
+                  BatchMutateContext.UpdateStatus updateStatus = context.getUpdateStatus(i);
+                  if (BatchMutateContext.UpdateStatus.UPDATE.equals(updateStatus)) {
+                      byte[] retVal = PInteger.INSTANCE.toBytes(1);
+                      Cell cell = PhoenixKeyValueUtil.newKeyValue(
+                              miniBatchOp.getOperation(i).getRow(),
+                              EMPTY_BYTE_ARRAY, EMPTY_BYTE_ARRAY, 0, retVal, 0,
+                              retVal.length);
+                      Result result = Result.create(new ArrayList<>(Arrays.asList(cell)));
+                      miniBatchOp.setOperationStatus(i,
+                              new OperationStatus(OperationStatusCode.SUCCESS, result));
+                  }
+              }
```

Comment:
Why do we need this logic again in `postBatchMutateIndispensably` ?

Reply:
Since HBase treats Put/Delete mutations as non-atomic mutations (unlike Increment/Append mutations), if we don't set Result object here in `postBatchMutateIndispensably()`, no Result object would stay around by the time mutation is executed till this point.

This is the reason why value `0` is provided with Result object early on (as part of `addOnDupMutationsToBatch()` as we short-circuit there by ignoring the mutation) whereas value `1` is provided with Result object here.

Reply:
I didn't understand your comment "No result object would stay around" ?. We are only setting Result with value `1` here. Result with value `0` are already set earlier and are still around right when this method is called ?

Reply:
> Result with value `0` are already set earlier and are still around right when this method is called ?

That's correct, because from HBase side, no remaining logic of batchMutate() gets executed as we short-circuit with value `0`.
Only when the whole batchMutate() execution goes through, Result object no longer stays around.
```
      batchOp.visitBatchOperations(true, finalLastIndexExclusive, (int i) -> {
        Mutation mutation = batchOp.getMutation(i);
        if (mutation instanceof Increment || mutation instanceof Append) {
          if (finalSuccess) {
            batchOp.retCodeDetails[i] =
              new OperationStatus(OperationStatusCode.SUCCESS, batchOp.results[i]);
          } else {
            batchOp.retCodeDetails[i] = OperationStatus.FAILURE;
          }
        } else {
          batchOp.retCodeDetails[i] =
            finalSuccess ? OperationStatus.SUCCESS : OperationStatus.FAILURE;
        }
        return true;
      });
```

Here, HBase sets `OperationStatus.SUCCESS` to `batchOp.retCodeDetails[i]` for Put/Delete mutations, leaving no Result object around with OperationStatus.
However, in case of ignore, the above logic is not executed, hence what we set with value `0` with `addOnDupMutationsToBatch()` stays around but for setting `1`, we need to set it with `postBatchMutateIndispensably()`.

Reply:
But for our purposes we don't need to send the Result with value `1`. Implicit `null` is good enough for us. If we don't set the Result for success a null will be returned which the client is already handling. Then we don't even need to maintain this map of update status. Also, since postBatchMutateIndispensably is also called for regular updates we will do a lookup in the map for every mutation in the batch and that lookup will not return anything but it is adding additional overhead.

Reply:
While that is true, i was thinking of two cases:

1. If there is a bug at server side, and the client starts receiving null Result object always, we would not know the difference b/ server returning null object by intention vs the buggy behavior causing override to what server is supposed to send.
2. In future, when we need to introduce JDBC API returning Result object with values, we anyways need to incorporate this logic.

How about we do this: keep the above code same, but wrap it within if condition: 
`if(context.hasAtomic && miniBatchOp.size() == 1)`

Reply:
After we introduce the check `if(context.hasAtomic && miniBatchOp.size() == 1)` what value is the map bringing ? Is it enough to justify maintaining a new map considering we are only using for batch size 1 ? I would recommend not maintaining the map at all for now. We can revisit this later in a subsequent PR on how to return Result object

Reply:
Getting rid of Map makes sense. Instead of `Map<Integer, UpdateStatus> updateStatusMap`, let's use `UpdateStatus updateStatus` and update it's value when `miniBatchOp.size() == 1` and operation is atomic.

Reply:
`updateStatusMap` is designed to replace the check of `miniBatchOp.getOperationStatus(i) == IGNORE`. If we get rid of Map, then we should change this check to something like `miniBatchOp.size() == 1 && updateStatus == IGNORE`, right?

Reply:
Yes @jinggou, I think that should work. Let's try to change it this way and see how it goes.

Reply:
Oh but we still need to preserve IGNORE status for batch atomic upsert, and without map it's tricky to know which index of the batch needs to be ignored.

Reply:
Oh but we still need to preserve IGNORE status for batch atomic upsert, and without map it's tricky to know which index of the batch needs to be ignored.
If not Map, we at least need ArrayList (or array) to get and set UpdateStatus for the indexed Mutation. Or how about we do this: we keep an array of indexes to be ignored i.e. 
`List<Integer> ignoreStatusIndexes = new ArrayList<>()`

Reply:
In-order to disntinguish between IGNORE and UPDATED can't we simply check if the Result object is null or not in the `OperationStatus` ? All those that will be ignored will have a non-null Result object.

Reply:
Good idea, this is worth trying!

Reply:
But now we have the Result object with value 0 for ignored and 1 for updated, right? Or do we want to go back to only sending the Result with value 0 for ignored, and implicit null is good enough to represent updated ?

Reply:
Discussed offline and my question is resolved.